File: packages/shared/src/utils/merge-deep.ts

/**
 * Recursively merges two or more objects.
 * Non-object inputs are skipped.
 * Arrays and primitive values are overwritten by later objects.
 *
 * @param objects - One or more objects to merge.
 * @returns A new object with merged properties.
 */
export function mergeDeep(...objects: any[]): any {
    // Return an empty object if no objects are provided.
    if (objects.length === 0) {
        return {};
    }

    const result: Record<string, any> = {};

    for (const obj of objects) {
        // Only process non-null objects that are not arrays.
        if (obj && typeof obj === "object" && !Array.isArray(obj)) {
            for (const key in obj) {
                if (Object.prototype.hasOwnProperty.call(obj, key)) {
                    const value = obj[key];

                    // If the value is an object (and not an array), merge recursively.
                    if (value && typeof value === "object" && !Array.isArray(value)) {
                        if (
                            result.hasOwnProperty(key) &&
                            result[key] &&
                            typeof result[key] === "object" &&
                            !Array.isArray(result[key])
                        ) {
                            result[key] = mergeDeep(result[key], value);
                        } else {
                            // Clone the object to avoid mutation.
                            result[key] = mergeDeep({}, value);
                        }
                    } else {
                        // For arrays and primitive values, override with the new value.
                        result[key] = value;
                    }
                }
            }
        }
    }

    return result;
}

File: packages/python_backend/app/utils/merge_deep.py
from typing import Any, Dict

def merge_deep(*objects: Dict[str, Any]) -> Dict[str, Any]:
    if not objects: return {}
    result: Dict[str, Any] = {}

    for obj in objects:
        if not isinstance(obj, dict): continue

        for key, value in obj.items():
            if isinstance(value, dict):
                if key in result and isinstance(result[key], dict):
                    result[key] = merge_deep(result[key], value)
                else:
                    result[key] = merge_deep({}, value)
            else:
                result[key] = value
    return result